test_conformance/api/test_api_consistency.cpp

```diff
@@ -0,0 +1,479 @@
+
+        // clGetMemObjectInfo, passing CL_MEM_USES_SVM_POINTER
+        // Returns CL_FALSE if no devices in the context associated with
+        // memobj support Shared Virtual Memory.
+        cl_bool usesSVMPointer;
+        error =
+            clGetMemObjectInfo(mem, CL_MEM_USES_SVM_POINTER,
+                               sizeof(usesSVMPointer), &usesSVMPointer, NULL);
+        test_error(error, "Unable to query CL_MEM_USES_SVM_POINTER");
+        if (usesSVMPointer != CL_FALSE)
+        {
+            log_error("CL_DEVICE_SVM_CAPABILITIES returned 0 but "
+                      "CL_MEM_USES_SVM_POINTER did not return CL_FALSE\n");
+            return TEST_FAIL;
+        }
+
+        // Check that the SVM APIs can be called.
+
+        // Returns NULL if no devices in context support Shared Virtual Memory.
+        void* ptr0 = clSVMAlloc(context, CL_MEM_READ_WRITE, allocSize, 0);
+        void* ptr1 = clSVMAlloc(context, CL_MEM_READ_WRITE, allocSize, 0);
+        if (ptr0 != NULL || ptr1 != NULL)
+        {
+            log_error("CL_DEVICE_SVM_CAPABILITIES returned 0 but clSVMAlloc "
+                      "returned a non-NULL value\n");
+            return TEST_FAIL;
+        }
+
+        // clEnqueueSVMFree, clEnqueueSVMMemcpy, clEnqueueSVMMemFill,
+        // clEnqueueSVMMap, clEnqueueSVMUnmap, clEnqueueSVMMigrateMem Returns
+        // CL_INVALID_OPERATION if the device associated with command_queue does
+        // not support Shared Virtual Memory.
+
+        cl_uint pattern = 0xAAAAAAAA;
+        error = clEnqueueSVMMemFill(queue, ptr0, &pattern, sizeof(pattern),
+                                    allocSize, 0, NULL, NULL);
+        test_failure_error(
+            error, CL_INVALID_OPERATION,
+            "CL_DEVICE_SVM_CAPABILITIES returned 0 but clEnqueueSVMMemFill did "
+            "not return CL_INVALID_OPERATION");
+
+        error = clEnqueueSVMMemcpy(queue, CL_TRUE, ptr1, ptr0, allocSize, 0,
+                                   NULL, NULL);
+        test_failure_error(
+            error, CL_INVALID_OPERATION,
+            "CL_DEVICE_SVM_CAPABILITIES returned 0 but "
+            "clEnqueueSVMMemcpy did not return CL_INVALID_OPERATION");
+
+        error = clEnqueueSVMMap(queue, CL_TRUE, CL_MAP_READ, ptr1, allocSize, 0,
+                                NULL, NULL);
+        test_failure_error(
+            error, CL_INVALID_OPERATION,
+            "CL_DEVICE_SVM_CAPABILITIES returned 0 but "
+            "clEnqueueSVMMap did not return CL_INVALID_OPERATION");
+
+        error = clEnqueueSVMUnmap(queue, ptr1, 0, NULL, NULL);
+        test_failure_error(
+            error, CL_INVALID_OPERATION,
+            "CL_DEVICE_SVM_CAPABILITIES returned 0 but "
+            "clEnqueueSVMUnmap did not return CL_INVALID_OPERATION");
+
+        error = clFinish(queue);
+        test_error(error, "Error calling clFinish after SVM operations");
+
+        clSVMFree(context, ptr0);
```

Comment:
This is only testing the case where `ptr0` is `NULL` which is guaranteed to be a no-op even when SVM is supported. Better to assign a value to `ptr0` beforehand.

Reply:
Well, we cannot allocate a valid SVM pointer when SVM isn't supported, so I'm not sure which value would be better. :smile:

We also have no way of knowing whether `clSVMFree` was a NOP or even if it executed successfully or returned an error, since it has no error code.  I think the best we can do is pass the value of `ptr0` that we have, which will be `NULL`, and be sure the call does not crash.

Reply:
My point was that it's guaranteed to be a no-op whether SVM is supported or not. This is not testing that clSVMFree always is a no-op (not that we can do better than "it doesn't crash" without a lot of effort). I'd just make up a value here (DEADBEEF or similar to make debugging more obvious if this ever triggers a crash).

Reply:
I'm hesitant to make this change since it's not clear to me that passing a bogus pointer to `clSVMFree` is any better than passing a `NULL` pointer.  How strongly do you feel about this?

Reply:
Well, passing `NULL` is specified to behave the same whether the feature is supported or not. There are bugs that will not be exposed by passing `NULL` but _may_ be exposed by passing a bogus value. I'd say this is unambiguously better.

Reply:
I'll just change this (and the call to `clEnqueueSVMFree` below) since it's a trivial change, but I will note that the current description for `clSVMFree` says (**emphasis** mine):

> **svm_pointer must be the value returned by a call to clSVMAlloc**. If a NULL pointer is passed in svm_pointer, no action occurs.

It doesn't say what happens if _svm_pointer_ is NOT a value returned by a call to clSVMAlloc, other than the NULL pointer case, but I suspect that many implementations are defensive and will validate that the input is a known SVM pointer before freeing it.  For these implementations, passing a bogus pointer will behave the same as passing a NULL pointer.

As an aide, it would be really nice to describe the expected behavior for `clSVMFree` when passed a bogus pointer in the case when SVM is supported - is this a NOP, undefined behavior, implementation-defined behavior, or something else?

Reply:
Yes but it doesn't really matter here since this code will only run when SVM is not supported and in that case the implementation should definitely not do anything with the pointer. That's what we're trying to test.

>As an aide, it would be really nice to describe the expected behavior for clSVMFree when passed a bogus pointer in the case when SVM is supported - is this a NOP, undefined behavior, implementation-defined behavior, or something else?

Agree it's worth specifying. I'd lean towards undefined or implementation-defined. This could be checked by a validation layer and would enable implementations to assume the pointer is correct.